src/overlays/actors/ovl_En_Fish/z_en_fish.c

```diff
@@ -47,73 +63,709 @@ static ColliderJntSphInit D_80A17040 = {
+    this->actor.shape.rot = this->actor.world.rot;
+
+    SkelAnime_Update(&this->skelAnime);
+
+    if (this->timer <= 0) {
+        Actor_Kill(&this->actor);
+        return;
+    }
+
+    if (this->timer <= 60) {
+        // Blink when about to disappear
+        if (frames & 4) {
+            this->actor.draw = EnFish_Draw;
+        } else {
+            this->actor.draw = NULL;
+        }
+    } else if (this->actor.bgCheckFlags & 0x20) { // In water
+        EnFish_Dropped_SetupSwimAway(this);
+    } else if (this->actor.bgCheckFlags & 1) { // On floor
+        EnFish_Dropped_SetupFlopOnGround(this);
+    }
+}
+
+void EnFish_Dropped_SetupSwimAway(EnFish* this) {
+    this->actor.home.pos = this->actor.world.pos;
+    this->actor.flags |= 0x10;
+    this->timer = 200;
+    this->actor.gravity = 0.0f;
+    this->actor.minVelocityY = 0.0f;
+    this->actor.shape.yOffset = 0.0f;
+    EnFish_SetInWaterAnimation(this);
+    this->actionFunc = EnFish_Dropped_SwimAway;
+    this->unk_250 = 5;
+}
+
+void EnFish_Dropped_SwimAway(EnFish* this, GlobalContext* globalCtx) {
+    s32 pad;
+
+    Math_SmoothStepToF(&this->actor.speedXZ, 2.8f, 0.1f, 0.4f, 0.0f);
+
+    // If touching wall or not in water, turn back and slow down for one frame.
+    if ((this->actor.bgCheckFlags & 8) || !(this->actor.bgCheckFlags & 0x20)) {
+        this->actor.home.rot.y = Math_Vec3f_Yaw(&this->actor.world.pos, &this->actor.home.pos);
+        this->actor.speedXZ *= 0.5f;
+    }
+
+    Math_StepToAngleS(&this->actor.world.rot.x, 0, 1500);
+    Math_StepToAngleS(&this->actor.world.rot.y, this->actor.home.rot.y, 3000);
+    Math_StepToAngleS(&this->actor.world.rot.z, 0, 1000);
+
+    this->actor.shape.rot = this->actor.world.rot;
+
+    // Raise if on a floor.
+    if (this->actor.bgCheckFlags & 1) {
+        Math_StepToF(&this->actor.world.pos.y, this->actor.home.pos.y - 4.0f, 2.0f);
+    } else {
+        Math_StepToF(&this->actor.world.pos.y, this->actor.home.pos.y - 10.0f, 2.0f);
+    }
+
+    // Shrink when close to disappearing.
+    if (this->timer < 100) {
+        Actor_SetScale(&this->actor, this->actor.scale.x * 0.982f);
+    }
+
+    this->skelAnime.playSpeed = CLAMP_MAX((this->actor.speedXZ * 1.5f) + 1.0f, 4.0f);
+    SkelAnime_Update(&this->skelAnime);
+
+    if (this->timer <= 0) {
+        Actor_Kill(&this->actor);
+    }
+}
+
+// Unique type functions
+
+void EnFish_Unique_SetupSwimIdle(EnFish* this) {
+    this->actor.gravity = 0.0f;
+    this->actor.minVelocityY = 0.0f;
+    this->timer = Rand_S16Offset(5, 35);
+    this->unk_250 = 0;
+    EnFish_SetInWaterAnimation(this);
+    this->actionFunc = EnFish_Unique_SwimIdle;
+}
+
+void EnFish_Unique_SwimIdle(EnFish* this, GlobalContext* globalCtx) {
+    static f32 speedStopping[] = { 0.0f, 0.04f, 0.09f };
+    static f32 speedMoving[] = { 0.5f, 0.1f, 0.15f };
+    f32 playSpeed;
+    u32 frames;
+    f32* speedType;
```

Comment:
Does this make sense?
```suggestion
    static Vec3f speedStopping = { 0.0f, 0.04f, 0.09f };
    static Vec3f speedMoving = { 0.5f, 0.1f, 0.15f };
    f32 playSpeed;
    u32 frames;
    Vec3f* speedType;
```

Reply:
I disagree with this: it's not treated like a Vec3f at all: it's the goal and the step size arguments.

Reply:
does a 2d array work?

Reply:
It does (Yay, another fake symbol!). Sadly `speedType = speedData[this->timer < 12];` doesn't work

Reply:
That said, I'm not actually sure it's clearer with a 2D array.

Reply:
its fine as is without 2d array imo. though id rename the `speedType` var. that makes it sound like the var denotes what the type is, when really its just the array of speed values. I think just `speed` is fine